### ISSUE TITLE
Inline form errors: Do we want to allow space for them? Or do we want to form to jump on errors?

### ISSUE THREAD

**Issue by lizsurette** [patternfly/patternfly-design#671]
The PatternFly-React team is currently implementing the latest Login designs. While talking through the best way to do inline errors, we have talked back and forth about the experience of the form either growing to allow for space to show the error or simply having space already built into the form for a less jumpy experience. You can see the initial discussion and some code examples [here](https://github.com/patternfly/patternfly-react/pull/362), but I think it would be great for the design team to think through this not only for the login form but for forms in general and how we want to handle errors inline.

**Comment by Laviro** [comment]
Thanks @lizsurette for opening this discussion.
Here are the examples as I promised, fading errors vs. collapsing errors.

**Fading** input errors with **collapsing** form error:
![38928ae5-d01a-45f3-y92e-abc11b67f656 1](https://user-images.githubusercontent.com/26363699/40706081-5c005dda-63f5-11e8-998a-ec8f90053955.gif)

**Collapsing errors:**
![0c75c7d8-2045-4568-y48b-e8499a6b7a83](https://user-images.githubusercontent.com/26363699/40705959-01b0f178-63f5-11e8-8d23-8fc5893803e9.gif)

**Comment by jgiardino** [comment]
In the same pf-react pr mentioned above, a [similar discussion](https://github.com/patternfly/patternfly-react/pull/362#pullrequestreview-124432635) has come up regarding the section of social media links. Here, the overall height of the login modal is being predefined to avoid resizing in height when the More link is clicked. Guidance on this aspect would be appreciated too.

**Comment by mcarrano** [comment]
@matthewcarleton Can you take a look at this issue and give an opinion? We discussed today in the PF Design Sync mtg and feel like this change would make sense for the login screen specifically, but not necessarily forms in general. Is it reasonable to apply that change in spacing to this pattern only or would that effect field spacing on forms in general?

**Comment by mcarrano** [comment]
@matthewcarleton and I discussed today. While we think there is merit to the proposal, making this change would potentially affect form layout across PatternFly and I don't think this is something we should consider for PatternFly 3. The other potential issue is that it's not always possible to know the length of error messages, so allowing for a single line message mat still result in problems when the message needs to wrap.

We recommend that this be revisited for PatternFly-Next where we can factor this into our overall form layout system.

**Comment by Laviro** [comment]
@mcarrano @matthewcarleton We can give the two options for developers to choose from,
and I have some solutions in mind:
- The component may decide by itself whether to fade or collapse by the line length/height.
- maybe even limit the provided error message length (like twitter).
- if we don't want to limit the message, we can put a -- [more...](url) -- link which will collapse the rest of the message.

**Comment by ohadlevy** [comment]
@mcarrano I just want to double check, are you suggesting these kind of issues are not going to be fixed for pf3? is there an official SLA for pf3? there are a lot of important projects that are using pf3 and do not plan to migrate to pf4 (as there is no upgrade plan yet), statements like these make me feel very worried and uncomfortable with the direction patternfly is headed.

**Comment by Laviro** [comment]
Hey everyone,
seems that the decision for not keeping a free space for the errors was wrong..
After I started to implement this component in the `Foreman` project,
the community really didn't liked the jumping errors, See: https://community.theforeman.org/t/our-new-login-page/11827

**Comment by Laviro** [comment]
Can I start working on a fix ?

**Comment by LHinson** [comment]
@lizsurette @Laviro we started to revisit this conversation on the PatternFly to determine what the best approach is. 

While I understand the negative reaction to the modal jumping in height, there is also an argument for the other side that this jump helps to reiterate to the user that something has changed and calls for their attention. 

Another potential concern is that making this change to the login form is inconsistent with how we approach forms elsewhere in PatternFly. 

Accounting for the extra space doesn’t look as nice up front and only really makes sense when someone does trigger an error (which is a subset of users vs all users that will see the initial impression). 

For the reasons above, we are not convinced on the PatternFly side at this time that this is the correct path, therefore we are not ready to merge anything to diverge from what is done today in PatternFly. 

One avenue could be to apply the adjustments specifically within your application while we explore alternative approaches to reporting field level errors for the login page to maintain a consistent layout.

**Comment by Laviro** [comment]
> this jump helps to reiterate to the user that something has changed and calls for their attention.

@LHinson  I agree on that for the form main error, as for the inline inputs error me & the community disagree.

>  inconsistent with how we approach forms elsewhere in PatternFly.

could you attach links to other forms with errors implementations ?

It doesn't have to be consistent if it's a rare case.. after all it isn't just a regular form,
it is a Login Page the won't be in use anywhere if we can't get people to like it.

I think patternfly as an opensource project should really care about the community feedback,
and this kind of approach won't work.

we need to think that maybe we were up to something here,
maybe inline form errors should behave this way.. and we need to fix it else where.

so please reconsider this..

In addition, I have another idea..
Currently we have the `LoginPagePattern` as a design pattern 
and we also have the `managed` pattern which interacts dynamically with the user.

maybe we can add the spaces to the managed version meanwhile
and leave the original spaces in the original design as this discussion grows.

Thanks.

**Comment by LHinson** [comment]
@Laviro I completely agree with the importance of community feedback and insight. This helps PatternFly to improve both our design and code and want to thank you for raising concerns and making these comments known to us. 

The intention of my earlier response was to highlight that the issue is not so black and white and that we aren't quite ready for a PR/change to PatternFly because of the conflicting perspectives. Here are the facts: 

- Foreman community members identified that they do not like the jumping. 
- PatternFly visuals design identified that accounting for the space has a negative impact to the visual design of the login (and the majority of what users see)
- CSS identified concerns with this approach since it violates CSS best practices of letting the content drive height and not setting fixed height for elements
- We currently do not have any usability data or research metrics 
- We do not have any feedback from users 

PatternFly must balance an iterative approach with project stability. Given the above points, I don't think we have a clear answer on how to proceed at this time. There may be a design solution that addresses all the concerns that we have not yet explored. 

Is this issue currently a blocker for proceeding with the new login design for Foreman?

**Comment by Laviro** [comment]
Thanks for explaining, 
as much as i don't like it,
I guess we will override the css.

**Comment by Rohoover** [comment]
@Laviro 

I would caution against the override. From reading the above thread - it seems like we are not dealing with an issue that is outright breaking the application. I would prefer overrides be reserved for times where the content is outright blocked, impeded, or functionally horrible. Even with the jump, it is not functionally horrible and is acting consistently. I would hate to engender an attitude that if we don't "like" something we simply over ride it. This will undoubtedly create greater inconsistency because of the individual contributor process of the upstreams.

At this point, I see this as a preference, since no one has enough data or enough user feedback to determine "the best way". I see many merits to what you proposed, but also understand the complexity of making behavior changes to forms. I know that you feel the login page isn't regular form, but I have to disagree with that particular point. In it's basic elements it is a form. 

I do agree with @ohadlevy that these issues need to be continually looked at as his statement that there are products that are not migrating anytime soon is factually correct. That said, we also need to prioritize the importance of any specific issue. 

- My 5 cents (I get 3 extra cents since I wrote 3 paragraphs)

**Comment by Laviro** [comment]
Thanks @Rohoover,
Though this is what I've understand from the above thread..

How about replacing the current input errors design
With something like PF4 LoginPage ?
![image](https://user-images.githubusercontent.com/26363699/49963896-e40f1f00-ff21-11e8-8e80-cde31fe10bf9.png)

With overlay on the alert icons which explains the error / warning?
My last 5 cent.. ;)

**Comment by LHinson** [comment]
@Laviro I really should have deferred to @Rohoover earlier regarding the UX of this specific application. I think she makes very valid points and would proceed with her recommendation. 

As for PF4 approach for errors, I'm glad you happened upon the PF4 designs. The adjustment in approach was influenced by the conversation that has happened to date. I do want to clarify that this design is still in progress and being evaluated since the messaging is initially hidden from the user.

As a heads up, today is my last day before heading out for maternity leave. I'm looping in @mcarrano @rachael-phillips for help to determine how we can move forward.

**Comment by mcarrano** [comment]
@Laviro @Rohoover I have discussed this with @mcoker in reference to how we might implement this for PatternFly 3 without making a breaking change and would like to make the following proposal. Rather than display authentication errors at the field level, we're recommending that you use the space above the login form to return an error. This would be equivalent to the PF4 example you show above and would look something like this: 

![image](https://user-images.githubusercontent.com/6619860/50293925-b60a6d00-0442-11e9-9839-a4bed764448b.png)

This approach is also preferable from a security standpoint as it's generally not good practice to identify whether the user ID or password is in error as it will make it easier to hack into the site. If you are OK with this approach, we can update the core CSS to allocate some space above the form to prevent the message from pushing other content down when it appears. This may require a corresponding update to the React component that I'm hoping you can contribute @Laviro . We can also document this recommended approach in the design documentation here: https://www.patternfly.org/pattern-library/application-framework/login-page/#design

Thoughts?

**Comment by Laviro** [comment]
Thanks @mcarrano, I like it,

As for empty input errors,
Can we decide that the submit button would be disabled until the inputs are filled ?

As for caps-lock,
Can we add warning Icon on the right-hand of the input ?

Just wanted to see if I understand,
when password is too short, and input is blured, 
should the main form error collapse and alert that it is too short?
Same goes for invalid email structure.

**Comment by mcarrano** [comment]
@Laviro here are my thoughts. Welcome other inputs.

> As for empty input errors,
> Can we decide that the submit button would be disabled until the inputs are filled ?

Yes, I think this makes sense.

> As for caps-lock,
> Can we add warning Icon on the right-hand of the input ?

I think this is a good approach to that. Would like to see what this looks like, however. Would you want that to then have a tooltip? I think it should. My only concern is consistency with what we are doing elsewhere. @Rohoover what do you think?

> Just wanted to see if I understand,
> when password is too short, and input is blured,

 > the main form error collapse and alert that it is too short?
> Same goes for invalid email structure.

My opinion is that these client-side errors should still be handled by inserting helper text below the field. This gives the user immediate feedback on the error when the input is blurred. I know that this would still cause a small shift in spacing, but I think this is a small price to pay for the localized feedback. Handling authentication errors always by providing a message above the form solves what I see as the main issue which was the Log In button jumping after it is clicked, which I agree is pretty annoying.

**Comment by Laviro** [comment]
Hey guys,
before starting to work on the fixes mentioned above, 
any other opinions or thoughts about what @mcarrano and I have mentioned ?

thanks

**Comment by Rohoover** [comment]
I think I would like to see what the capslock solution being proposed looks like as well.

**Comment by MariaAga** [comment]
@Laviro and I were thinking of something like this:
![screencast-localhost-6006-2019 01 13-16-16-59](https://user-images.githubusercontent.com/30431079/51086401-3aed5b80-174f-11e9-8eb6-b741550d8b1f.gif)

**Comment by Laviro** [comment]
@MariaAga also started to work on a [PR](https://github.com/patternfly/patternfly-react/pull/1170) to change the way we present warnings on the input.

**Comment by Laviro** [comment]
@mcarrano @Rohoover could you have a look on the progress that was made in the PRs ? Thanks :)

**Comment by andybraren** [comment]
I'm in agreement with @mcarrano's thoughts above, and the proposed Caps Lock hint implementation looks good to me. I wonder if using a standard [Caps Lock icon](https://icons8.com/icon/31852/caps-lock-on) would be recognizable enough to not require the tooltip to be open by default, but we'd need to research that or do usability testing to confirm. From what I can tell FontAwesome doesn't currently have that icon either, unfortunately.

One possible alternative to immediately showing helper text below the field when blurred would be to use HTML5 `minlength="3"` and `type="email"` attributes on the form inputs ([examples here](https://www.w3schools.com/html/html_form_attributes.asp)). If you want this to work in IE11 as well, [`pattern`](https://caniuse.com/#feat=input-pattern) should be capable of doing the same, or you could just rely on the existing error messages for that particular browser.

These attributes would display a tooltip and automatically focus the first invalid field after the user clicks submit, avoiding the jumpiness of doing so on blur. I haven't investigated it fully, so if it doesn't suit your needs then I'd once again agree with @mcarrano's recommendation of sticking with localized syntax help below the invalid input field, either on blur or after clicking submit.

Here's what an invalid email address looks like using `type="email"`:

<img width="752" alt="2019-01-18 11 05 39" src="https://user-images.githubusercontent.com/9122899/51398489-b6fdef80-1b11-11e9-94b7-abead6a3cfd6.png">

**Comment by Laviro** [comment]
Thanks @andybraren,
I prefer using our own warnings implementation,
though this could be created by passing the HTML attributes through props.

Take a look on Maria's work mentioned above, I think it is ready to be implemented.

**Comment by jgiardino** [comment]
I looked at the example in https://github.com/patternfly/patternfly-react/pull/1170. I noticed that the when error message disappears, the form doesn't jump but transitions nicely to it's new size. Is this behavior new with this PR? That transition feels really nice. 👍 

I just wanted to remind everyone that with PF3, we're asking that any [new contributions are keyboard accessible](https://codeguide.patternfly.org/#html-accessibility). This means that any information that is accessible with a mouse is also accessible using only a keyboard. To make the message accessible to the keyboard-only users, we can display the tooltip on focus as well as hover.

**Comment by Laviro** [comment]
Thanks @jgiardino the transition was there earlier :)
Yes, we can add the accesibility you mentioned, @MariaAga 
but first I'd like to know if this designs are accepted @mcarrano @Rohoover 
And whether this discussion can be resolved :)

**Comment by Rohoover** [comment]
I'm a little concerned that these designs are going to have cross-application implications that we're not fully aware of. We're also creating inconsistencies for forms as a general mechanism in PatternFly. 

I'm also a little concerned about supplying action level feedback for passwords from a security standpoint. Wouldn't the correct feedback for caps lock on be to provide the proper error handling at the top? I sort of feel like we're trying to solve for a niche problem that is already solved through the current mechanism.

**Comment by Rohoover** [comment]
All - @Laviro and I were able to meet and chat about this topic this AM. We had a productive conversation and I want to summarize that here:

- We are going to opt to remove iconography from the fields as well as inline error messaging for the login. 
- We will place error messages at the top as was suggested earlier by @mcarrano which is one of the suggested design options and should relieve the jumping issue.
- I expressed PF's standpoint that the Login is still a form and changes just to that break with the consistency amongst other forms.
- We talked about the benefits of his design proposal but feel like it would be best to tackle this amongst all forms and not break it out as just a login issue. Perhaps we can pursue this as a larger change to forms with input from the larger PF community. We understand this may take time.

Overall this is really a great discussion topic and hopefully some of these suggestions can be vetted and incorporated as a larger effort. 

@Laviro Did I miss anything?
